#16 - Indoor directions

NicholasNagy opened the pull request:
I made some refactorings as a result of the feedback from SonarQube.

Some of these changes include updating todo's, refactoring complex methods and renaming methods and classes.
I also made a slight update to the design of the arrow for the indoor navigation.

We also added some more tests to cover these new areas of the code.

NicholasNagy committed:
Massive refactoring of handle navigation method

This refactoring was to remove the code smell which was that the cognitive complexity was at 40, there was some duplicate code, uselessly nested if statements
- Separating the functionality of the buttons from the generation of directions
- the logic that initializes the generation of directions are now put into their own methods

NicholasNagy committed:
Refactor generate directions method

I refactored the generate directions into several methods since it was large and gave it a more appropriate name

NicholasNagy committed:
Refactored readsvg template method

Refactored readsvg template method to reduce the cognitive complexity from 26 to 15.

In order to do this, I made a separate function that collects an element and runs the appropriate lambda function

NicholasNagy committed:
Adding explanation to empty lambda functions

I added comments in 2 lambda functions to explain why they are empty as they are a code smell

NicholasNagy committed:
Refactored attribute extraction

Attribute extraction was made in an overly complicated way, so I decided to rewrite the code in a way that made more sense and was easier to understand.

NicholasNagy committed:
Fixed todo in getStartAndEnd

There was a todo in this method, and after taking a careful look, I realized that this portion of the code did not do anything, so I removed it.

NicholasNagy committed:
Refactored point generation

I refactored/cleanuped up the generate points method as its cognitive complexity was too high.

I did this by extracting the clean up functionality out of the method

NicholasNagy committed:
Optimizing point generation for indoor paths

I cut down on the code that meticulously puts code in corners, and added more points blindly since the shortest path algorithm does not slow down with many points like with dijkstra.

JeremiahTee committed:
Merge branch 'master' into indoorDirections

NicholasNagy committed:
Merge remote-tracking branch 'origin/master' into indoorDirections

NicholasNagy committed:
Deleting patch

Removing patch that neeham had pushed.

NicholasNagy committed:
Merge branch 'indoorDirections' of https://github.com/RobertBeaudenon/SOEN390-CampusGuideMap into indoorDirections

JeremiahTee committed:
Refactor name variables in ExplorePlace

For Kotlin standards

JeremiahTee committed:
Merge remote-tracking branch 'origin/indoorDirections' into indoorDirections

JeremiahTee committed:
Revert "Merge remote-tracking branch 'origin/indoorDirections' into indoorDirections"

This reverts commit e460aca4b1f80ee4ff73684a580df2c60b7dee70, reversing
changes made to e1840d159b48e7d2039f3d44dffb11606e334e1f.

JeremiahTee committed:
Revert "Refactored readsvg template method"

This reverts commit 293f5496b70dd1981eab930583f5028961a3f4a8.

NicholasNagy committed:
Revert "Refactored readsvg template method"

This reverts commit 293f5496b70dd1981eab930583f5028961a3f4a8.

NicholasNagy committed:
Adding ability to choose indoor transportation method

- Allowing users to choose their indoor transportation method using the settings menu

- Currently when no available options are available for a building, it displays a toast.

JeremiahTee committed:
UI test for SettingsFragment, final coverage of 84%

Updated/fixed unit tests and added a UI test for Settings
The final statement coverage of the project is 84% with 63% branch coverage

JeremiahTee committed:
Merge remote-tracking branch 'origin/indoorDirections' into indoorDirections

NicholasNagy committed:
Fixing bug in mb-s1

I fixed a bug where mb-s1 would not updates its room numbers

NicholasNagy committed:
Merge branch 'indoorDirections' of https://github.com/RobertBeaudenon/SOEN390-CampusGuideMap into indoorDirections

NicholasNagy committed:
Fixing empty nav step

I fixed an empty nav step that was induced by adding an additional step when initializing the chain of navhandlers

NicholasNagy committed:
Last minute refactoring method names

I refactored method names to better reflect what they do, and to respect the naming convention.

NicholasNagy committed:
Improving the style of the arrows

I improved the styling of the drawn arrow when display a path from start to end for indoor navigation

NicholasNagy committed:
Removing commented out line of code

This line of code is no longer used since we don't use the closest point to do anything anymore, we use all the points generated now to get a better spread of the area.

KhaledBadran commented on app/src/main/java/com/droidhats/campuscompass/views/FloorFragment.kt:
```diff
@@ -387,34 +417,59 @@ class FloorFragment : Fragment(), NavigationView.OnNavigationItemSelectedListene
+     * The functionality for this is to launch the appropriate directions to generate the next
+     * step.
+     *
+     * @param startAndEnd
+     * @param goingUp
+     * @param building
+     * @param doneButton
+     * @param prevArrowButton
```
Can we add a description of each of the parameters?

KhaledBadran commented on app/src/main/java/com/droidhats/campuscompass/views/FloorFragment.kt:
```diff
@@ -477,18 +514,28 @@ class FloorFragment : Fragment(), NavigationView.OnNavigationItemSelectedListene
+    /**
+     * Generate the directions for the first step.
+     *
+     * @param startAndEnd
+     * @param goingUp
+     * @param building
```
Can we give a description to the used parameters?

KhaledBadran commented on app/src/main/java/com/droidhats/campuscompass/views/FloorFragment.kt:
```diff
@@ -531,26 +578,52 @@ class FloorFragment : Fragment(), NavigationView.OnNavigationItemSelectedListene
+    /**
+     * Generate the directions for the second step.
+     *
+     * @param startAndEnd
+     * @param goingUp
+     * @param building
```
Similar case, can we add a description for the used parameter?

KhaledBadran commented on app/src/main/java/com/droidhats/campuscompass/views/FloorFragment.kt:
```diff
@@ -564,72 +637,110 @@ class FloorFragment : Fragment(), NavigationView.OnNavigationItemSelectedListene
+        setDirections(startPos, endPos, mapProcessor)
+    }
+
+
```
Can we remove the extra space?

KhaledBadran commented on app/src/main/java/com/droidhats/campuscompass/views/FloorFragment.kt:
```diff
@@ -564,72 +637,110 @@ class FloorFragment : Fragment(), NavigationView.OnNavigationItemSelectedListene
+    /**
+     * Setting the directions on the image view
+     *
+     * @param start
+     * @param end
+     * @param mapProcessor
```
I think it would be nice to add descriptions to those parameters.

KhaledBadran commented on app/src/main/java/com/droidhats/mapprocessor/AStar.kt:
```diff
@@ -117,37 +117,18 @@ fun findNearestPoint(mapElement: MapElement, pathElements: List<Vertex>): Vertex
 /**
  * Takes 2 map elements and returns the shortest path in between them given the list of path elements
+ * as an arrow
  * @param start element
  * @param end element
  * @param list of vertices
  * @return string
```
Can we add more details to the parameters of this method. The return is not needed since the method description covers it.

KhaledBadran commented on app/src/main/java/com/droidhats/mapprocessor/AStar.kt:
```diff
@@ -117,37 +117,18 @@ fun findNearestPoint(mapElement: MapElement, pathElements: List<Vertex>): Vertex
-        }
-    }
+    val path = aStar(startVertex, endVertex, start, end, pathElements)
+    if (path != null) return path
```
Can we add parentheses to this if statement, or possibly move the return statement to a new line with proper indentation.

KhaledBadran commented on app/src/main/java/com/droidhats/mapprocessor/AStar.kt:
```diff
@@ -181,6 +161,46 @@ fun A_Star(start: MapElement, end: MapElement, pathElements: MutableList<Vertex>
 
+/**
+ * A* algorithm for getting the shortest path quickly
+ * @param startVertex
+ * @param endVertex
+ * @param start
+ * @param end
+ * @param pathElements
```
Is it possible to add a description to those parameters?

KhaledBadran reviewed (changes requested):
Really nice updates to the code. The new changes contribute significantly to the code quality. I especially like the addition of exception handling to the various scenarios of indoor navigation. It makes the app much more robust and allows it to fail gracefully without breaking. 

Other than some minor changes in the code that I've left in my comments, the changes are solid and working perfectly 😁 👍

Amanda-Kolopa reviewed (commented):
I agree with many of the comments left in Khaled's review, we need to fix those comments up, aside from that, I tested a lot of the functionalities that were touched in this PR, and everything seems to be working fine.

However, there are a few enhancements that might be necessary. When navigating between floors using the elevator, the arrow brings you to the top portion of the map, on the other side of the elevator, might need to modify that.

NicholasNagy committed:
Updating elevator locations

I updated the elevator locations so that the nearest path point can make sense when navigating to these locations.

NicholasNagy committed:
Fixing styling in tests

Since I added styling to the path and circle variables, some of the unit tests were failing, so I updated them with the new values.

NicholasNagy committed:
Elaborated on dokka comments

As requested by Khaled, I updated the dokka comments so that the parameter are more informative and easier to understand instead of just displaying the parameter name.

NicholasNagy commented:
Hey guys, so I fixed the comments that you guys left for me, especially relating to that elevator, here are the before and after pics:
![Screenshot_20200414_215638_com droidhats campuscompass 1977](https://user-images.githubusercontent.com/40705372/79294878-488b1700-7ea5-11ea-8c4b-a73524d92548.jpg)
![Screenshot_20200414_222922_com droidhats campuscompass 1976](https://user-images.githubusercontent.com/40705372/79294885-4cb73480-7ea5-11ea-8097-5af367b0f693.jpg)
